### ISSUE TITLE
Dynamically contol the number of parallel downloads

### ISSUE THREAD

**Issue by jplatte** [serpent-os/tools#377]
Especially on slow connections, it would be nice to have fewer DL jobs in parallel. On the other hand, on faster connections (and when there are lots of small packages to download), it probably makes sense to have more parallel downloads.

**Comment by jplatte** [comment]
I just spent way too much time thinking about this and coming up with a scheme for how this could be done. It's 100% based on my intuition and no hard data or prior art, but here goes 😄 

- Start downloading with 3 parallel jobs
- Every time a download is finished or every 100ms after the last measurement, measure (calculate) the average response body transmission speed over the last 5s (summed across all jobs from that time)
  - Based on the measured tx speed, calculate a desired number of jobs
  - If the desired number of jobs is higher than the current number of jobs, start new jobs until the desired number is reached or the end of the download queue is reached

For the tx speed to desired number of jobs, playing around with numbers until it looked nice lead me to the following scheme:

tx speed (kb/s)  | jobs | note: bandwidth per job (tx speed/jobs)
------|-------|------
< 384 | 1 | 0 - 384
≥ 384   | 2            | 192 - 256
≥ 512   | 3            | 170 - 227
≥ 682   | 4            | 170 - 227
≥ 910   | 5            | 182 - 242
≥ 1213  | 6            | 202 - 269
≥ 1618  | 7            | 213 - 308
≥ 2157  | 8            | 269 - 359
≥ 2876  | 9            | 319 - 426
≥ 3835  | 10           | 383 - 511
≥ 5114  | 11           | 464 - 619
≥ 6816  | 12           | 568 - 757
≥ 9091  | 13           | 699 - 932
≥ 12122 | 14           | 865 - 1154
≥ 16163 | 15           | 1077 - 1436
≥ 21551 | 16           | 1346+

tx speed increases ~1.333333-fold per row, a logarithmic scale like that felt better than a flat "+N kb/s means +1 job" and in particular 1.2 felt too "fast" (too many job bumps as tx speed increases), while sqrt(2) felt too slow.

All of these numbers could of course be tweaked.

If I'm not the only one who thinks this looks good, I could probably spend some time on an implementation soon-ish (no promises though).

**Comment by joebonrichie** [comment]
I spend some time some months ago looking at this, I wish to reformat my disk so i'm gonna dump the patch i had here.

The biggest issue with it iirc it that it worked in batches starting from 3 jobs and waited for the batch to complete before adding or reducing number of download jobs, so it could take a while to saturate a fast connection.

<details>
<summary>download-jobs-batches.patch</summary>

```diff
diff --git a/moss/src/client/mod.rs b/moss/src/client/mod.rs
index c2ff39d..52bb0bf 100644
--- a/moss/src/client/mod.rs
+++ b/moss/src/client/mod.rs
@@ -11,9 +11,14 @@
 
 use std::{
     borrow::Borrow,
+    collections::VecDeque,
     fmt, io,
     os::{fd::RawFd, unix::fs::symlink},
     path::{Path, PathBuf},
+    sync::{
+        atomic::{AtomicU64, AtomicUsize},
+        Arc,
+    },
     time::{Duration, Instant},
 };
 
@@ -465,20 +470,38 @@ impl Client {
 
         let unpacking_in_progress = cache::UnpackingInProgress::default();
 
-        // Download and unpack each package
-        let cached = stream::iter(packages)
-            .map(|package| async {
+        let mut current_concurrency = 3;
+        // Track download performance
+        let mut speed_tracker = DownloadSpeedTracker::new();
+
+        // Convert packages to a Vec for processing
+        let mut remaining_packages: Vec<_> = packages.iter().collect();
+
+        // Collect all results for DB operations at the end
+        let mut all_cached_results = Vec::with_capacity(packages.len());
+
+        while !remaining_packages.is_empty() {
+            // Take a batch of packages based on current concurrency
+            let batch_size = current_concurrency.min(remaining_packages.len());
+            let batch: Vec<_> = remaining_packages.drain(..batch_size).collect();
+
+            let start_time = Instant::now();
+            let total_bytes = Arc::new(AtomicUsize::new(0));
+            let non_cached_count = Arc::new(AtomicUsize::new(0));
+
+            // Process batch with futures
+            let mut futures = Vec::with_capacity(batch.len());
+
+            for package in batch {
                 let package: &Package = package.borrow();
+                let package_name = package.meta.name.to_string();
+                let expected_size = package.meta.download_size.unwrap_or_default();
 
-                // Setup the progress bar and set as downloading
-                let progress_bar = multi_progress.insert_before(
+                // Create progress bar for this download
+                let download_bar = multi_progress.insert_before(
                     &total_progress,
-                    ProgressBar::new(package.meta.download_size.unwrap_or_default())
-                        .with_message(format!(
-                            "{} {}",
-                            "Downloading".blue(),
-                            package.meta.name.to_string().bold(),
-                        ))
+                    ProgressBar::new(expected_size)
+                        .with_message(format!("{} {}", "Downloading".blue(), package_name.clone().bold(),))
                         .with_style(
                             ProgressStyle::with_template(
                                 " {spinner} |{percent:>3}%| {wide_msg} {binary_bytes_per_sec:>.dim} ",
@@ -487,62 +510,139 @@ impl Client {
                             .tick_chars("--=≡■≡=--"),
                         ),
                 );
-                progress_bar.enable_steady_tick(Duration::from_millis(150));
-
-                // Download and update progress
-                let download = cache::fetch(&package.meta, &self.installation, |progress| {
-                    progress_bar.inc(progress.delta);
-                })
-                .await?;
-                let is_cached = download.was_cached;
-
-                // Move rest of blocking code to threadpool
+                download_bar.enable_steady_tick(Duration::from_millis(150));
 
+                // Set up progress tracking
+                let total_bytes = total_bytes.clone();
+                let non_cached_count = non_cached_count.clone();
                 let multi_progress = multi_progress.clone();
                 let total_progress = total_progress.clone();
                 let unpacking_in_progress = unpacking_in_progress.clone();
-                let package = (*package).clone();
 
-                runtime::unblock(move || {
-                    let package_name = package.meta.name.to_string();
+                // Create the future
+                let future = async move {
+                    // Track download bytes
+                    let bytes_counter = Arc::new(AtomicU64::new(0));
+                    let counter_clone = bytes_counter.clone();
+
+                    // Download with progress
+                    let download_result = cache::fetch(&package.meta, &self.installation, |progress| {
+                        download_bar.inc(progress.delta);
+                        counter_clone.fetch_add(progress.delta, std::sync::atomic::Ordering::Relaxed);
+                    })
+                    .await;
+
+                    // Handle download errors
+                    let download = match download_result {
+                        Ok(d) => d,
+                        Err(e) => {
+                            download_bar.finish_with_message(format!("Failed to download {}", package_name.clone()));
+                            multi_progress.remove(&download_bar);
+                            return Err(e.into());
+                        }
+                    };
 
-                    // Set progress to unpacking
-                    progress_bar.set_message(format!("{} {}", "Unpacking".yellow(), package_name.clone().bold()));
-                    progress_bar.set_length(1000);
-                    progress_bar.set_position(0);
+                    // Update stats for concurrency calculation
+                    let downloaded_bytes = bytes_counter.load(std::sync::atomic::Ordering::Relaxed);
+                    let is_cached = download.was_cached;
 
-                    // Unpack and update progress
-                    let unpacked = download.unpack(unpacking_in_progress.clone(), {
-                        let progress_bar = progress_bar.clone();
+                    if !is_cached {
+                        total_bytes.fetch_add(downloaded_bytes as usize, std::sync::atomic::Ordering::Relaxed);
+                        non_cached_count.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+                    }
 
-                        move |progress| {
-                            progress_bar.set_position((progress.pct() * 1000.0) as u64);
-                        }
-                    })?;
+                    // Set up for unpacking
+                    download_bar.set_message(format!("{} {}", "Unpacking".yellow(), package_name.clone().bold()));
+                    download_bar.set_length(1000);
+                    download_bar.set_position(0);
 
-                    // Remove this progress bar
-                    progress_bar.finish();
-                    multi_progress.remove(&progress_bar);
+                    // Create a channel for progress updates
+                    let (progress_tx, progress_rx) = std::sync::mpsc::channel();
 
-                    let cached_tag = is_cached
-                        .then_some(format!("{}", " (cached)".dim()))
-                        .unwrap_or_default();
+                    // Thread to update progress bar
+                    let progress_thread = std::thread::spawn(move || {
+                        while let Ok(pct) = progress_rx.recv() {
+                            download_bar.set_position((pct * 1000.0) as u64);
+                        }
+                        // Clean up progress bar when done
+                        download_bar.finish();
+                        //multi_progress.borrow().remove(&download_bar);
+                    });
+
+                    // Unpack on threadpool
+                    let unpack_result = runtime::unblock({
+                        let package = (*package).clone();
+                        //let package_name = package_name.clone();
+                        //let multi_progress = multi_progress.clone();
+                        let total_progress = total_progress.clone();
+                        let unpacking_in_progress = unpacking_in_progress.clone();
+                        let is_cached = is_cached;
+                        let progress_tx = progress_tx.clone();
+
+                        move || {
+                            // Create progress callback that sends through channel
+                            let progress_callback = move |progress: cache::Progress| {
+                                let _ = progress_tx.send(progress.pct());
+                            };
+
+                            // Unpack with progress updates
+                            let unpacked = download.unpack(unpacking_in_progress, progress_callback)?;
+
+                            let cached_tag = is_cached
+                                .then_some(format!("{}", " (cached)".dim()))
+                                .unwrap_or_default();
+
+                            // Report installation success
+                            multi_progress.clone().suspend(|| {
+                                println!("{} {}{cached_tag}", "Installed".green(), package_name.clone().bold())
+                            });
+
+                            total_progress.inc(1);
+
+                            Ok::<_, Error>((package, unpacked, is_cached))
+                        }
+                    })
+                    .await?;
+
+                    // Close progress channel to terminate thread
+                    drop(progress_tx);
+                    // Wait for thread to finish
+                    let _ = progress_thread.join();
 
-                    // Write installed line
-                    multi_progress
-                        .suspend(|| println!("{} {}{cached_tag}", "Installed".green(), package_name.clone().bold()));
+                    Ok(unpack_result)
+                };
 
-                    // Inc total progress by 1
-                    total_progress.inc(1);
+                futures.push(future);
+            }
 
-                    Ok((package, unpacked)) as Result<(Package, cache::UnpackedAsset), Error>
-                })
+            // Wait for all futures in this batch to complete
+            let batch_results = futures_util::future::join_all(futures)
                 .await
-            })
-            // Use max network concurrency since we download files here
-            .buffer_unordered(environment::MAX_NETWORK_CONCURRENCY)
-            .try_collect::<Vec<_>>()
-            .await?;
+                .into_iter()
+                .collect::<Result<Vec<_>, Error>>()?;
+
+            // Add to our results
+            all_cached_results.extend(batch_results);
+
+            // Adjust concurrency based on performance
+            let bytes = total_bytes.load(std::sync::atomic::Ordering::Relaxed);
+            let non_cached = non_cached_count.load(std::sync::atomic::Ordering::Relaxed);
+
+            if non_cached > 0 && bytes > 0 {
+                let elapsed = start_time.elapsed();
+                if elapsed.as_millis() > 0 {
+                    let speed_mbps = (bytes as f64 / elapsed.as_secs_f64()) / (1024.0 * 1024.0);
+                    speed_tracker.add_measurement(speed_mbps);
+
+                    // Update concurrency for next batch
+                    current_concurrency = speed_tracker.suggest_concurrency(environment::MAX_NETWORK_CONCURRENCY);
+
+                    // Optionally log the adjustment
+                    // println!("Download speed: {:.2} MB/s, adjusting concurrency to: {}",
+                    //         speed_mbps, current_concurrency);
+                }
+            }
+        }
 
         // Add layouts & packages to DBs
         runtime::unblock({
@@ -555,7 +655,7 @@ impl Client {
                 total_progress.tick();
 
                 // Add layouts
-                layout_db.batch_add(cached.iter().flat_map(|(p, u)| {
+                layout_db.batch_add(all_cached_results.iter().flat_map(|(p, u, _)| {
                     u.payloads
                         .iter()
                         .flat_map(PayloadKind::layout)
@@ -567,7 +667,7 @@ impl Client {
                 total_progress.set_message("Storing DB packages");
 
                 // Add packages
-                install_db.batch_add(cached.into_iter().map(|(p, _)| (p.id, p.meta)).collect())?;
+                install_db.batch_add(all_cached_results.into_iter().map(|(p, _, _)| (p.id, p.meta)).collect())?;
 
                 total_progress.inc(1);
 
@@ -994,6 +1094,57 @@ impl BlitStats {
     }
 }
 
+struct DownloadSpeedTracker {
+    speeds: VecDeque<f64>,  // Recent speed measurements in MB/s
+    window_size: usize,     // Number of measurements to keep
+    min_concurrency: usize, // Minimum concurrency allowed
+    trend_threshold: f64,   // Speed change percentage to trigger adjustment
+}
+
+impl DownloadSpeedTracker {
+    fn new() -> Self {
+        Self {
+            speeds: VecDeque::with_capacity(5),
+            window_size: 5,
+            min_concurrency: 2,
+            trend_threshold: 0.15, // 15% change threshold
+        }
+    }
+
+    fn add_measurement(&mut self, speed_mbps: f64) {
+        self.speeds.push_back(speed_mbps);
+        if self.speeds.len() > self.window_size {
+            self.speeds.remove(0);
+        }
+    }
+
+    fn suggest_concurrency(&self, max_concurrency: usize) -> usize {
+        if self.speeds.len() < 2 {
+            return self.min_concurrency.max(3); // Default early value
+        }
+
+        // Calculate the trend
+        let prev_avg = self.speeds.iter().take(self.speeds.len() - 1).sum::<f64>() / (self.speeds.len() - 1) as f64;
+        let current_speed = self.speeds[self.speeds.len() - 1];
+
+        // Determine if speed is improving or declining
+        let ratio = current_speed / prev_avg;
+
+        let current_concurrency = self.speeds.len().min(max_concurrency);
+
+        if ratio > (1.0 + self.trend_threshold) {
+            // Speed is improving, try higher concurrency
+            return (current_concurrency + 1).min(max_concurrency);
+        } else if ratio < (1.0 - self.trend_threshold) {
+            // Speed is declining, reduce concurrency
+            return (current_concurrency - 1).max(self.min_concurrency);
+        }
+
+        // Speed is stable, stick with current concurrency
+        current_concurrency
+    }
+}
+
 /// Client-relevant error mapping type
 #[derive(Debug, Error)]
 pub enum Error {
```
</details>

**Comment by ermo** [comment]
@jplatte If you are still interested in this, I think you and Joey should begin looking into this...?

**Comment by jplatte** [comment]
Yeah, I'm still interested but I need to find some focus time for this, which is hard.
